runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java

```diff
@@ -145,26 +153,32 @@ public Token appendCheckpoint() {
      * This API should not be directly invoked.
      *
      *  @param snapshotTimestamp snapshot at which the checkpoint is taken.
+     *  @param streamTail tail of the stream to checkpoint at snapshot time.
      */
     @VisibleForTesting
-    public Token appendCheckpoint(Token snapshotTimestamp) {
+    public Token appendCheckpoint(Token snapshotTimestamp, Long streamTail) {
         long start = System.currentTimeMillis();
 
-        // If snapshotTimestamp (stream tail) is a negative value
+        // If stream tail is a negative value
         // we need to enforce a hole to ensure the start log address
         // is not negative (as MultiCheckpointWriter will select the min of all checkpoints)
         // Having a negative value will prevent trimming from happening.
-        if (!Address.isAddress(snapshotTimestamp.getSequence())) {
+        if (!Address.isAddress(streamTail)) {
             snapshotTimestamp = forceHole();
+            streamTail = snapshotTimestamp.getSequence();
         }
 
-        try {
-            rt.getAddressSpaceView().read(snapshotTimestamp.getSequence());
-        } catch (TrimmedException te) {
-            // Stream Tail was trimmed, we can just ignore this trim. This means
-            // no other update has been done to this stream since trim.
-            log.info("appendCheckpointOnSnapshot: stream tail at {} has been trimmed. Checkpoint on same state" +
-                    "as previous checkpoint.", snapshotTimestamp);
+        // This read attempts to ensure no future sequencer regression by materializing the global tail, i.e.,
+        // preventing reassignment of tokens on sequencer failover (for the non-materialized space).
+        rt.getAddressSpaceView().read(snapshotTimestamp.getSequence());
```

Comment:
I think it will be cleaner if the stream marking code is refactored into a seperate method. Maybe introduce a method called `markStream` which returns a "stable" token.